lib/usr/clib/msgchan/msgchan.h

```diff
@@ -0,0 +1,146 @@
+
+/**
+ * @file
+ * Message Channels
+ *
+ * Create a message channel using two lockless rings to communicate between two threads.
+ *
+ * Message channels are similar to pipes in Linux and other platforms, but does not support
+ * message passing between processes.
+ *
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define MC_NO_CHILD_CREATE \
+    0x80000000 /**< If set in mc_create() flags then child will not be created */
+
+typedef void msgchan_t; /**< Opaque msgchan structure pointer */
+
+/**
+ * @brief Create a message channel
+ *
+ * Calling mc_create() with an existing channel name will create a child
+ * channel attached to the parent channel.
+ *
+ * @param name
+ *   The name of the message channel
+ * @param sz
+ *   The number of entries in the lockless ring for each direction.
+ * @param flags
+ *   The cne_ring_t flags for SP/SC or MP/MC type flags, look at cne_ring_create()
+ *   Defaults to (RING_F_MP_ENQ | RING_F_MC_DEQ) if flags is zero. Use the flags
+ *   (RING_F_SP_ENQ | RING_F_SC_DEQ);
+ *   Or in the bit MC_NO_CHILD_CREATE to not allow creating a child, NULL will be returned.
+ * @return
+ *   The pointer to the msgchan structure or NULL on error
+ */
+CNDP_API msgchan_t *mc_create(const char *name, int sz, uint32_t flags);
+
+/**
+ * @brief Destroy the message channel and free resources.
+ *
+ * @param mc
+ *   The msgchan structure pointer to destroy
+ * @return
+ *   N/A
+ */
+CNDP_API void mc_destroy(msgchan_t *mc);
+
+/**
+ * @brief Send object messages to the other end of the channel
+ *
+ * @param mc
+ *   The message channel structure pointer
+ * @param objs
+ *   An array of void *objects to send
+ * @param count
+ *   The number of entries in the objs array.
+ * @return
+ *   -1 on error or number of objects sent.
+ */
+CNDP_API int mc_send(msgchan_t *mc, void **objs, int count);
+
+/**
+ * @brief Receive message routine from other end of the channel
+ *
+ * @param mc
+ *   The message channel structure pointer
+ * @param objs
+ *   An array of objects pointers to place the received objects pointers
+ * @param count
+ *   The number of entries in the objs array.
+ * @param msec
+ *   Number of milliseconds to wait for data, if return without waiting.
+ * @return
+ *   -1 on error or number of objects
+ */
+CNDP_API int mc_recv(msgchan_t *mc, void **objs, int count, uint64_t msec);
+
+/**
+ * @brief Lookup a message channel by name parent only lookup
+ *
+ * @param name
+ *   The name of the message channel to find, which is for parent channels
+ * @return
+ *   NULL if not found, otherwise the message channel pointer
+ */
+CNDP_API msgchan_t *mc_lookup(const char *name);
+
+/**
+ * @brief Return the name string for the msgchan_t pointer
+ *
+ * @param mc
+ *   The message channel structure pointer
+ * @return
+ *   NULL if invalid pointer or string to message channel name
+ */
+CNDP_API const char *mc_name(msgchan_t *mc);
+
+/**
+ * @brief Return size and free space in the server/Client rings.
```

Comment:
not sure if we want to stick to the producer/consumer nomenclature vs the server/client... what do you think?